src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -1953,6 +1953,101 @@ void M2(object o)
             );
         }
 
+        [Fact, WorkItem(32879, "https://github.com/dotnet/roslyn/issues/32879")]
+        public void ThrowExpressionInNullCoalescingOperator()
+        {
+            var source = @"
+class C
+{
+    string Field = string.Empty;
+    void M(C? c)
+    {
+        _ = c ?? throw new System.ArgumentNullException(nameof(c));
+        c.ToString();
+    }
+    void M2(C? c)
+    {
+        _ = c?.Field ?? throw new System.ArgumentNullException(nameof(c));
+        c.ToString();
+        c.Field.ToString();
+    }
+    void M3(C? c)
+    {
+        _ = c ?? throw new System.ArgumentNullException(c.ToString()); // 1
+    }
+    void M4(string? s)
+    {
+        _ = s ?? s.ToString(); // 2
+        s.ToString();
+    }
+    void M5(string s)
+    {
+        _ = s ?? s.ToString(); // 3
+    }
+}
+";
+
+            // We should probably warn of possible de-reference in alternative case of (3)
```

Comment:
>probably [](start = 25, length = 8)

Nit: not probably, definitely. If we're in that case, there is _definitely_ a null reference exception getting thrown.

Reply:
I tend to agree, however the linked issue is marked as a design issue. It is queued on the agenda with some related questions about reachability.

Reply:
Update to include the `string~` case as well: after the `??` operator, we should definitely infer that `s` is possibly null, not oblivious.

---
In reply to: [253553088](https://github.com/dotnet/roslyn/pull/32991#discussion_r253553088) [](ancestors = 253553088)

Reply:
@jcouv this wasn't addressed. It's also fine to update the linked bug if you want to avoid having another commit.

---
In reply to: [253981194](https://github.com/dotnet/roslyn/pull/32991#discussion_r253981194) [](ancestors = 253981194,253553088)

Reply:
@333fred I thought I had addressed your feedback by adding:
```C#
#nullable disable
    void M6(string s)
    {
#nullable enable
        _ = s ?? s.ToString();
    }
```
Is that not what you had in mind?